[UI/UX] Expand navigation patterns with more task pane specific advice (#5617)
OfficeDev/office-js-docs-pr: docs/design/navigation-patterns.md

    <li><a href="#settings">Settings</a></li>
  </ul>
</nav>
```

## Decision tree: Choosing a navigation pattern

Use this guide to select the appropriate navigation pattern for your add-in.

1. **How many top-level sections does your add-in have?**
   - One to two sections: No persistent navigation needed; use ribbon commands and simple content.
   - Two to four sections: **Tab Bar** (if equal importance) or **Command Bar with menu** (if hierarchical).
   - Four to seven sections: **Command Bar with hamburger menu**.
   - Eight or more sections: Reconsider information architecture and consolidate.

1. **Do users frequently switch between sections?**
   - Yes: **Tab Bar** (keep all options visible).
   - No: **Command Bar with menu** (save space).

1. **Does your add-in have multistep workflows?**
   - Yes: Use **Back Button** and **Progress Indicator** for linear flows.
   - No: Use direct navigation (tabs or menu).

1. **Does your add-in have more than one point of entry from the ribbon?**
   - Yes: Use function commands with a [shared runtime](../develop/configure-your-add-in-to-use-a-shared-runtime.md) to **deep link** from ribbon to specific task pane views.

>    - No: Use a standard `ShowTaskpane` action in your manifest.

Reviewer: ```suggestion
   - No: Use a standard `openPage` (unified manifest for Microsoft 365) or `ShowTaskpane` (add-in only manifest) action in your manifest.
```

Office Add-ins present unique navigation challenges due to their constrained task pane width, integration with the Office ribbon, and multiplatform requirements. Effective navigation ensures users can access features easily while maintaining context with their Office document.

This article covers navigation patterns within task panes and how to coordinate navigation between the Office ribbon and your add-in UI.

## Unique navigation challenges for Office Add-ins

When designing navigation for Office Add-ins, consider the following constraints.

- **Narrow task pane width**: Task panes range from 320 px (Outlook on the web) to 350 px (Excel), limiting horizontal navigation space. For detailed task pane dimensions and variants, see [Task panes in Office Add-ins](task-pane-add-ins.md).
- **Personality menu obstruction**: The Office personality menu can block top-right UI elements. For exact dimensions and positioning, see [Task panes in Office Add-ins](task-pane-add-ins.md#personality-menu).
- **Ribbon integration**: Navigation must coordinate between ribbon commands and task pane state. See [Add-in commands](add-in-commands.md) for ribbon integration details.

> - **Multiplatform variations**: Navigation UI might need to adapt for Web, Windows, Mac, and mobile platforms.

Reviewer: ```suggestion
- **Multiplatform variations**: Navigation UI might need to adapt for web, Windows, Mac, and mobile platforms.
```

## Ribbon and task pane coordination

Office Add-ins have two primary UI surfaces: the Office ribbon (via [add-in commands](add-in-commands.md)) and the task pane. Plan how these surfaces work together.

### When to use ribbon commands vs. task pane navigation

Use ribbon commands for:

- **Primary entry points** - Major features users access directly from Office (Insert, Format, Analyze).
- **Context-specific actions** - Commands that act on selected content (Format selection, Insert chart).
- **Quick actions** - Single-click operations that don't require extra input.
- **Feature discovery** - Making key capabilities visible without opening the task pane.

Use task pane navigation for:

- **Multistep workflows** - Processes requiring multiple inputs or decisions.
- **Settings and configuration** - Options that don't fit in a single dialog.
- **Content browsing** - Galleries, lists, or catalogs users explore.
- **Persistent state** - Features users keep open while working in their document.

### Coordinating ribbon and task pane state

> When a user clicks a ribbon command that opens the task pane, deep link to the relevant view. Use `Office.addin.showAsTaskpane()` to programmatically show the task pane, and then navigate to the appropriate view based on which ribbon command was clicked.

Reviewer: ```suggestion
When a user selects a ribbon command that opens the task pane, deep link to the relevant view. Use `Office.addin.showAsTaskpane()` to programmatically show the task pane, and then navigate to the appropriate view based on which ribbon command was clicked.
```

> When a user clicks a ribbon command that opens the task pane, deep link to the relevant view. Use `Office.addin.showAsTaskpane()` to programmatically show the task pane, and then navigate to the appropriate view based on which ribbon command was clicked.

Author: You're right. I'll revise that.

}
```

If you have multiple ribbon commands, each should open the task pane to a specific, relevant view. Avoid having different commands open the same view or the default home screen.

> [!NOTE]
> For complete details on programmatically showing and hiding task panes, see [Show or hide the task pane of your Office Add-in](../develop/show-hide-add-in.md).

### Using contextual tabs with task pane navigation

For Excel add-ins, [contextual tabs](contextual-tabs.md) provide ribbon commands that appear only when specific conditions are met (for example, when a chart is selected). Use contextual tabs to:

- Reduce ribbon clutter by showing commands only when relevant.
- Provide quick actions on selected content without changing the task pane state.
- Coordinate with the task pane - contextual tab commands can trigger task pane navigation changes.

**Example:** A data visualization add-in might show a contextual tab when a chart is selected, with a "Format Chart" button that opens the task pane to chart formatting options.

## Task pane navigation patterns

The following patterns work within the constrained space of an Office task pane.

### Command bar

> The [CommandBar](https://react.fluentui.dev/?path=/docs/components-commandbar--default) is a surface at the top of the task pane that houses commands and navigation controls. It typically appears immediately below the add-in name.

Reviewer: ```suggestion
The command bar is a surface at the top of the task pane that houses commands and navigation controls. It typically appears immediately below the add-in name.
```

Paged linked to doesn't exist.

Use recognizable icons with tooltips to make icon-only commands that save space. Make 44x44 px touch targets so your add-in is more accessible.

> [!IMPORTANT]
> Position right-aligned commands at least 40 px from the top-right corner to avoid the Office personality menu.

### Tab bar

The tab bar navigates with buttons that use vertically stacked text and icons. Tabs appear at the top of the task pane and switch between major sections.

:::image type="content" source="../images/add-in-tab-bar.png" alt-text="Illustration showing a tab bar within an Office desktop application task pane. This example shows a tab bar immediately below the add-in name with 'Home', 'Settings', 'Favorites', and 'Account' tabs.":::

Use a tab bar when:

- Your add-in has 2-4 primary sections of equal importance.
- Users frequently switch between sections.
- Each section has distinct, non-overlapping content.

Because task panes have a narrow width, design tab bars with the following considerations.

- **Width**: Limit to 3-4 tabs maximum in task panes.
  - 2 tabs: ~160-185px each.
  - 3 tabs: ~106-123px each.

>   - 4 tabs: ~80-92px each.

Author: For the record, I left the numerals in because of the old grammar rule that you don't mix numerals and words in the same sentence. However, I agree with your suggestions that this form looks better.

Consider using [Fluent UI React Nav](https://react.fluentui.dev/?path=/docs/components-nav--default) components to retain the Office look and feel. Highlight the active or selected navigation item and support keyboard navigation (Tab, Arrow keys) for accessibility.

## Multi-step workflows and wizards

For linear, multistep processes (onboarding, setup, configuration), use these patterns:

### Stepped progress indicator

Show users where they are in a multistep process with a visual progress indicator. This pattern is useful for onboarding, setup wizards, and configuration flows.

Use a simple horizontal progress bar (20-40 px height) to save vertical space. Place **Previous** and **Next** buttons in a fixed footer so they're always accessible.

> Limit wizards to 3-5 steps maximum. Allow users to skip optional steps and provide a **Cancel** or **Exit** option to return to the main add-in. Save progress automatically so users can resume if they close the task pane.

Reviewer: ```suggestion
Limit wizards to three to five steps. Allow users to skip optional steps and provide a **Cancel** or **Exit** option to return to the main add-in. Save progress automatically so users can resume if they close the task pane.
```

  Office.context.document.settings.set('currentView', view);
  Office.context.document.settings.saveAsync();

  // You can also use localStorage for user-specific state.
  localStorage.setItem('lastView', view);
}

// Restore the view when the task pane reopens.
Office.onReady(() => {
  const savedView = Office.context.document.settings.get('currentView')
                    || localStorage.getItem('lastView')
                    || 'home';
  navigateTo(savedView);
});
```

> [!NOTE]
> To learn more about persisting add-in state and settings, including best practices for using `Office.context.document.settings`, `localStorage`, and storage partitioning, see [Persist add-in state and settings](../develop/persisting-add-in-state-and-settings.md).

### Deep linking from ribbon commands

Enable ribbon buttons to open the task pane to specific views. This pattern is especially important when you have multiple ribbon commands that should each open different sections of your task pane.

```javascript
// Call this function from your ribbon command handler (function command).
function openSettings(event) {
  localStorage.setItem('openToView', 'settings');
  Office.addin.showAsTaskpane();

>   event.completed();

Author: Adds excess complication to the sample.

}

// In your task pane startup, check for the target view and navigate accordingly.
Office.onReady(() => {
  const targetView = localStorage.getItem('navigationTarget');
  if (targetView) {
    navigateTo(targetView);
    localStorage.removeItem('navigationTarget');
  }
});
```

> For more information about creating function commands in your manifest, see [Create add-in commands](../develop/create-addin-commands.md).

Reviewer: This section feels redundant since deep linking is already mentioned earlier in the article. Consider removing it.

## Multi-container navigation flows

Office Add-ins can use multiple UI containers. Plan ahead to ensure smooth navigation for the user.

### Task pane to dialog navigation

Dialogs provide focused, modal experiences that overlay the Office document. Use dialogs for:

- Authentication and sign-in flows.
- Critical confirmations.
- Forms that require the user's full attention.
- Content that needs more space than the task pane provides.

When you open a dialog from a task pane, use `Office.context.ui.displayDialogAsync()`. Use message passing to communicate between the dialog and task pane.

Some best practices for task pane to dialog navigation include:

- **Clear triggers**: Use explicit buttons, such as "Sign In", to open dialogs.
- **Return path**: Always provide a way to close the dialog and return to the task pane.
- **State synchronization**: Update task pane UI when the dialog closes.
- **Error handling**: Handle when the dialog closes via X button or Escape key.

For complete details on using dialogs, including code examples and API reference, see [Use the Office dialog API in Office Add-ins](../develop/dialog-api-in-office-add-ins.md).

## Content add-in navigation

> [Content add-ins](content-add-ins.md) are embedded directly in Excel, PowerPoint, or Word documents and have different navigation constraints than task panes. All UI is within the embedded content boundary, and users can resize the content add-in. Use simple navigation patterns such as single-page layouts or minimal tabs. For complex content add-ins, consider using a companion task pane for settings.

Reviewer: Per [Content Office Add-ins ](https://learn.microsoft.com/office/dev/add-ins/design/content-add-ins), content add-ins are only supported in Excel and PowerPoint. Additionally, I don't think we can implement a task pane in a content add-in.

```suggestion
[Content add-ins](content-add-ins.md) are embedded directly in Excel or PowerPoint documents and have different navigation constraints than task panes. All UI is within the embedded content boundary, and users can resize the content add-in. Use simple navigation patterns such as single-page layouts or minimal tabs.
```

Author: Fair enough. For some reason, I thought you could have a singular add-in with both, but you're right.

## Platform-specific navigation considerations

### Office on the web

Office on the web has the narrowest task panes. Test your navigation at 320 px width (Outlook on the web) to ensure it works on the most constrained platform. Users might click the browser back button, so use `history.pushState()` to handle this gracefully. Consider responsive breakpoints for phones and tablets accessing Office on the web.

### Office on Windows and Mac

Office on Windows and Mac supports slightly wider task panes, up to 350 px. Test keyboard navigation by using Tab, arrow keys, and standard shortcuts (such as Ctrl+F for search) to make sure your navigation is accessible. Reserve the top-right 40x40 px space to avoid the personality menu.

> ### Outlook mobile (iOS/Android)

Reviewer: ```suggestion
### Outlook on mobile devices (iOS/Android)
```

All navigation must be accessible by using a keyboard with Tab, arrow keys, and Enter. Provide a clear visual indication of which element has focus. Use proper ARIA labels, including `role="navigation"`, `aria-label`, and `aria-current="page"` for screen reader support. Provide a "Skip to content" link for screen reader users and ensure navigation appears early in the tab order, before main content.

The following example shows accessible navigation markup.

```html
<nav role="navigation" aria-label="Main navigation">
  <a href="#main-content" class="skip-link">Skip to content</a>
  <ul>
    <li><a href="#home" aria-current="page">Home</a></li>
    <li><a href="#settings">Settings</a></li>
  </ul>
</nav>
```

## Decision tree: Choosing a navigation pattern

Use this guide to select the appropriate navigation pattern for your add-in.

1. **How many top-level sections does your add-in have?**
   - 1-2 sections: No persistent navigation needed; use ribbon commands and simple content.
   - 2-4 sections: **Tab Bar** (if equal importance) or **Command Bar with menu** (if hierarchical).
   - 4-7 sections: **Command Bar with hamburger menu** or **Vertical Nav** (icon-only).

>    - 8+ sections: Reconsider information architecture and consolidate.

Reviewer: ```suggestion
   - One to two sections: No persistent navigation needed; use ribbon commands and simple content.
   - Two to four sections: **Tab Bar** (if equal importance) or **Command Bar with menu** (if hierarchical).
   - Four to seven sections: **Command Bar with hamburger menu**.
   - Eight or more sections: Reconsider information architecture and consolidate.
```

1. **Do users frequently switch between sections?**
   - Yes: **Tab Bar** (keep all options visible).
   - No: **Command Bar with menu** (save space).

1. **Does your add-in have multistep workflows?**
   - Yes: Use **Back Button** and **Progress Indicator** for linear flows.
   - No: Use direct navigation (tabs or menu).

1. **Does your add-in have more than one point of entry from the ribbon?**
   - Yes: Plan **deep linking** from ribbon to specific task pane views.

>    - No: Task pane can be independent of the ribbon.

Reviewer: I'm not sure I fully understand this statement. Does this refer to a task pane only requiring a single button on the ribbon, with all add-in operations handled within the task pane?

>    - Consider **Contextual Tabs** (Excel only) for context-specific commands.

Reviewer: ```suggestion
   - **Excel only**: Consider **Contextual Tabs** for context-specific commands.
```
